### PR TITLE
[JEP-305] [JENKINS-51308] Incrementalify Metrics plugin

### PR DESCRIPTION
**Pull request by batmat**
[JENKINS-51308](https://issues.jenkins-ci.org/browse/JENKINS-51308)

@stephenc cc @jglick

### CODE REVIEW

**Comment by batmat** [comment]
Followup of https://github.com/jenkins-infra/repository-permissions-updater/pull/693

**Comment by batmat** [comment]
\o/ https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/metrics/

File: pom.xml [hunk 1]
```diff
@@ -29,11 +29,11 @@
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>2.37</version>
+    <version>3.10</version>
   </parent>
 
   <artifactId>metrics</artifactId>
-  <version>3.1.2.13-SNAPSHOT</version>
+  <version>${revision}${changelist}</version>
```
Comment (stephenc) [inline pom.xml]: AIUI this is incompatible with the changes to the pom made by the release plugin

Comment (batmat) [reply pom.xml]: Fixed @stephenc, thanks for the heads-up.

Also, it worked: https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/metrics/4.0.2.0-rc159.21b416d8612c/

File: pom.xml [hunk 2]
```diff
@@ -33,7 +33,7 @@
   </parent>
 
   <artifactId>metrics</artifactId>
-  <version>4.0.2.0-SNAPSHOT</version>
+  <version>${revision}${changelist}</version>
```
Comment (stephenc) [inline pom.xml]: As I said before , this is incompatible with using the release plugin.

I did a local `mvn release:prepare -DpushChanges=false` and this is what I got:

```
$ git log -n 3
commit 6bd1162977ff30464f5cef32b4859cbdfea0e2f3 (HEAD -> batmat-JENKINS-51308)
Author: Stephen Connolly <stephen.alan.connolly@gmail.com>
Date:   Thu Jun 21 15:16:16 2018 +0100

    [maven-release-plugin] prepare for next development iteration

commit f2ff01e30b97fbca38c649ee738bd20590f55d10 (tag: metrics-4.0.2.2)
Author: Stephen Connolly <stephen.alan.connolly@gmail.com>
Date:   Thu Jun 21 15:16:16 2018 +0100

    [maven-release-plugin] prepare release metrics-4.0.2.2

commit 5f86aa4fe63f3e2454abbac199567ae7fed7e585
Merge: 34c6bea 21b416d
Author: Stephen Connolly <stephen.alan.connolly@gmail.com>
Date:   Thu Jun 21 15:13:10 2018 +0100

    Merge branch 'JENKINS-51308' of https://github.com/batmat/metrics-plugin into batmat-JENKINS-51308

$ git diff HEAD^^
diff --git a/pom.xml b/pom.xml
index 9c3dad0..90a977f 100644
--- a/pom.xml
+++ b/pom.xml
@@ -33,7 +33,7 @@
   </parent>
 
   <artifactId>metrics</artifactId>
-  <version>${revision}${changelist}</version>
+  <version>4.0.2.3-SNAPSHOT</version>
   <packaging>hpi</packaging>
 
   <name>Metrics Plugin</name>
```

So *every* time somebody does a real release this change will be reverted by the release plugin.

I am not merging a PR for Incrementals if is blocks cutting actual releases or if it adds manual steps to the release process.

File: pom.xml (hunk 1 above)
Comment (stephenc) [reply pom.xml]: You didn't fix it... the release plugin is still going to trample over this every time we cut a real release

File: pom.xml (hunk 2 above)
Comment (dwnusbaum) [reply pom.xml]: @stephenc Yes you have to manually run `mvn incrementals:reincrementalify` after doing a release if you want to reenable incremental release ([Documentation](https://github.com/jenkinsci/incrementals-tools#running-maven-releases)). There is a ticket to do it automatically as part of the release process [JENKINS-51658](https://issues.jenkins-ci.org/browse/JENKINS-51658).

**Comment by stephenc** [comment]
Closing until [JENKINS-51658](https://issues.jenkins-ci.org/browse/JENKINS-51658) is fixed.

**Comment by rtyler** [comment]
That ticket has been fixed, so this can be re-opened I believe.

**Comment by batmat** [comment]
@stephenc I updated to latest incrementals + latest parent pom to get auto-reincrementalification. (Introduced in [3.18+](https://github.com/jenkinsci/plugin-pom/blob/master/CHANGELOG.md#318)).

Care to have another review?

Thanks!

**Comment by batmat** [comment]
@stephenc @dwnusbaum care to review and/or merge this? Thanks

**Comment by dwnusbaum** [comment]
Given that Stephen's concern was addressed in [version 3.18 of the parent pom](https://github.com/jenkinsci/plugin-pom/blob/master/CHANGELOG.md#318) which has been picked up by this PR, I am going to go ahead and merge this.

**Comment by batmat** [comment]
http://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/metrics/4.0.2.3-rc229.d9b0eb3c576d/

**Pull request merged**